forest/src/paramfetch/mod.rs

```diff
@@ -0,0 +1,239 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use async_std::{fs, sync::Arc, task};
+use blake2b_simd::State as Blake2b;
+use core::time::Duration;
+use fil_types::SectorSize;
+use log::{info, warn};
+use pbr::{MultiBar, ProgressBar, Units};
+use reqwest::{blocking::Client, header, Proxy, Url};
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::error::Error as StdError;
+use std::fs::File;
+use std::io::{self, copy, prelude::*, ErrorKind, Stdout};
+use std::path::{Path, PathBuf};
+
+const GATEWAY: &str = "https://proofs.filecoin.io/ipfs/";
+const PARAM_DIR: &str = "/var/tmp/filecoin-proof-parameters";
+const DIR_ENV: &str = "FIL_PROOFS_PARAMETER_CACHE";
+const GATEWAY_ENV: &str = "IPFS_GATEWAY";
+const TRUST_PARAMS_ENV: &str = "TRUST_PARAMS";
+const DEFAULT_PARAMETERS: &str = include_str!("parameters.json");
+
+/// Sector size options for fetching.
+pub enum SectorSizeOpt {
+    /// All keys and proofs gen params
+    All,
+    /// Only verification params
+    Keys,
+    /// All keys and proofs gen params for a given size
+    Size(SectorSize),
+}
+
+type ParameterMap = HashMap<String, ParameterData>;
+
+#[derive(Debug, Deserialize, Serialize)]
+struct ParameterData {
+    cid: String,
+    digest: String,
+    sector_size: u64,
+}
+
+#[inline]
+fn param_dir() -> String {
+    std::env::var(DIR_ENV).unwrap_or_else(|_| PARAM_DIR.to_owned())
+}
+
+/// Get proofs parameters and all verification keys for a given sector size given
+/// a param JSON manifest.
+pub async fn get_params(
+    param_json: &str,
+    storage_size: SectorSizeOpt,
+    is_verbose: bool,
+) -> Result<(), Box<dyn StdError>> {
+    fs::create_dir_all(param_dir()).await?;
+
+    let params: ParameterMap = serde_json::from_str(param_json)?;
+    let mut tasks = Vec::with_capacity(params.len());
+
+    let mb = if is_verbose {
+        Some(Arc::new(MultiBar::new()))
+    } else {
+        None
+    };
+
+    for (name, info) in params {
+        match storage_size {
+            SectorSizeOpt::All => (),
+            SectorSizeOpt::Keys => {
+                if name.ends_with(".params") {
+                    continue;
+                }
+            }
+            SectorSizeOpt::Size(size) => {
+                if size as u64 != info.sector_size && name.ends_with(".params") {
+                    continue;
+                }
+            }
+        }
+
+        let cmb = mb.clone();
+        tasks.push(task::spawn(async move {
+            if let Err(e) = fetch_verify_params(&name, &info, cmb).await {
+                warn!("Error in validating params {}", e);
+            }
+        }));
+    }
+
+    if let Some(multi_bar) = mb {
+        let cmb = multi_bar.clone();
+        let (mb_send, mut mb_rx) = futures::channel::oneshot::channel();
```

Comment:
just curious why we are using this channel vs async-std channel?

Reply:
yeah no real reason other than just quicker to use oneshot channel for what I needed, I'll switch though (don't think it really matters though)